Backport `trace frames with np.ndarray` (#110512)

ev-br opened the pull request:
Backport https://github.com/pytorch/pytorch/pull/110512 to 2.1.1.

This one depends on https://github.com/pytorch/pytorch/pull/112027

Also, contains the cherry-picked fix from https://github.com/pytorch/pytorch/pull/112175

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ev-br committed:
Backport `trace frames with np.ndarray` (#110512)

Fixes #109604

Resubmit gh-109715 + several skips and small fixes to make tests pass.

The main fix here is by @ysiraichi : previously, dynamo did not resume tracing numpy ndarrays after a graph break.
While at it, fix several small issues Yukio's fix uncovers:

- graph break gracefully on numpy dtypes which do not map to torch.dtypes (uint16 etc)
- recognize array scalars in dynamo, treat them as 0D ndarrays
- make sure that iterating over torch.ndarray generates arrays not bare tensors

Pull Request resolved: https://github.com/pytorch/pytorch/pull/110512
Approved by: https://github.com/lezcano

lezcano commented:
It looks you'll also need to add the PR that fixes the `as_array(list[str])` issue

malfet commented on torch/_dynamo/variables/builder.py:
```diff
@@ -476,6 +476,9 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.ID_MATCH),
             )
+        elif isinstance(value, np.generic):
```
This will crash if numpy is not installed, wouldn't it?

malfet reviewed (changes requested):
Looks like this will crash if numpy is not installed, will submit a fix soon

malfet committed:
[Dynamo] Do not crash if numpy is not installed

s/isinstance(value, np.generic)/np is not None and isinstance(value, np.generic)/

huydhn commented:
@pytorchbot rebase -b release/2.1

lezcano commented:
Note that this PR depends on https://github.com/pytorch/pytorch/pull/112027

ev-br commented:
Missed the 2.1.1 window, so closing.
